### PR TITLE
Implemented getReachableRoleNames method, since I saw a lot of calls …

### PR DESCRIPTION
**Pull request by 0xPr0xy**
| Q | A |
| --- | --- |
| Branch? | "master" |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | none |
| License | MIT |
| Doc PR | none |

…to getReachableRoles, followed by an array map to get the role names

### CODE REVIEW

**Comment by yceruto** [comment]
@0xPr0xy this can be already reached through `getReachableRole` ?

**Comment by 0xPr0xy** [comment]
@yceruto Yes you are right, this will expose a Role[].
However, I see a number of occurrences in the symfony2 source, that point to a need for this method.
SecurityDataCollector.php L:102 and L:103:
`array_map(function (RoleInterface $role) { return $role->getRole();}, $assignedRoles)`
ExpressionVoter.php L:106:
`array_map(function ($role) { return $role->getRole(); }, $roles)`

And a number of occurrences in our own application bundles that we would like to refactor to get the reachableRoleNames essentially making the array map unneccesary

What are your thoughts?

File: src/Symfony/Component/Security/Core/Role/RoleHierarchy.php [hunk 1]
```diff
@@ -52,6 +52,29 @@ public function getReachableRoles(array $roles)
...
+            if (isset($this->map[$roleName]) === false) {
+                continue;
+            }
+
+            foreach ($this->map[$roleName] as $subRoleName) {
+                $reachableRoleNames[] = $subRoleName;
+            }
+        }
+
+        return $reachableRoleNames;
```
Comment (yceruto) [inline src/Symfony/Component/Security/Core/Role/RoleHierarchy.php]: why not simplify:
```
public function getReachableRoleNames(array $roles)
{
    return array_map(
        function (RoleInterface $role) {
            return $role->getRole();
        },
        $this->getReachableRoles($roles)
    );
}
```

**Comment by 0xPr0xy** [comment]
I guess that would be more elegant, also I guess I could also apply array_key_exists instead of isset in the getReachableRoles() method. and then also implement getRoleNames in TokenInterface and implementing classes, and refactor the underlying code to not use the array map anymore but use the appropriate getters. What do you think?

**Comment by robfrawley** [comment]
@0xPr0xy Unless you need to distinguish between a *non-existent key* and a *key who's value is `null`*, then sticking with `isset` over `array_key_exists` has performance benefits. That said, this is a super-micro-optimization. ;-)

**Comment by ro0NL** [comment]
What about `getReachableRoles(array $roles, $namesOnly = false)`? Or map where needed (like now), ie. im not sure how big a deal this is.

However the 3 instances you mentioned cant be simply refactored to `getReachableRolesNames` as `Token::getRoles` is also influenced.

- https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php#L84
- https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Security/Core/Authorization/Voter/ExpressionVoter.php#L82

**Comment by ro0NL** [comment]
@0xPr0xy what about a `Symfony\Component\Security\Core\Role\Role::getNames(array $roles)` utility?

**Comment by fabpot** [comment]
Closing as the short version as suggested in the comments seems good enough. The second reason being that we cannot add a new method in an interface as this would be a BC break (getting around would mean adding a new interface, which I think is not worth it for such a small addition).

**Pull request closed without merging**